framework/src/constraints/AutomaticMortarGeneration.C

```diff
@@ -1933,24 +2034,47 @@ AutomaticMortarGeneration::projectPrimaryNodesSinglePair(
 }
 
 void
-AutomaticMortarGeneration::writeNodalNormalsToFile()
+AutomaticMortarGeneration::writeGeometryToFile()
 {
-  // Must call compute_nodal_normals() first!
-  if (_secondary_node_to_nodal_normal.empty())
-    mooseError("No entries found in the secondary node -> nodal normal map.");
-
-  // Note: I seem to remember an issue with creating a second
-  // EquationSystems with the same Mesh, but this code seems to work
-  // OK currently.
-  EquationSystems nodal_normals_es(_mesh);
-  ExplicitSystem & nodal_normals_system =
-      nodal_normals_es.add_system<ExplicitSystem>("nodal_normals");
-  auto nnx_var_num = nodal_normals_system.add_variable("nodal_normal_x", FEType(FIRST, LAGRANGE)),
-       nny_var_num = nodal_normals_system.add_variable("nodal_normal_y", FEType(FIRST, LAGRANGE));
-  nodal_normals_es.init();
-
-  const DofMap & dof_map = nodal_normals_system.get_dof_map();
-  std::vector<dof_id_type> dof_indices_nnx, dof_indices_nny;
+  if (!_debug)
+    return;
+
+  // Must call compute_nodal_geometry first!
+  if (_secondary_node_to_nodal_normal.empty() || _secondary_node_to_hh_nodal_tangents.empty())
+    mooseError("No entries found in the secondary node -> nodal geometry map.");
+
+  auto & problem = _app.feProblem();
+  auto & subproblem = _on_displaced ? static_cast<SubProblem &>(*problem.getDisplacedProblem())
+                                    : static_cast<SubProblem &>(problem);
+  auto & nodal_normals_es = subproblem.es();
+
+  if (!_nodal_normals_system)
+  {
+    _nodal_normals_system = &nodal_normals_es.template add_system<ExplicitSystem>("nodal_normals");
+    _nnx_var_num = _nodal_normals_system->add_variable("nodal_normal_x", FEType(FIRST, LAGRANGE)),
+    _nny_var_num = _nodal_normals_system->add_variable("nodal_normal_y", FEType(FIRST, LAGRANGE));
+    _nnz_var_num = _nodal_normals_system->add_variable("nodal_normal_z", FEType(FIRST, LAGRANGE));
+
+    _t1x_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_x", FEType(FIRST, LAGRANGE)),
+    _t1y_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_y", FEType(FIRST, LAGRANGE));
+    _t1z_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_z", FEType(FIRST, LAGRANGE));
+
+    _t2x_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_x", FEType(FIRST, LAGRANGE)),
+    _t2y_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_y", FEType(FIRST, LAGRANGE));
+    _t2z_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_z", FEType(FIRST, LAGRANGE));
+    nodal_normals_es.reinit();
```

Comment:
I know I wrote this, but would we ever want to output the non-vertex normals?

Reply:
The result of the normal vector interpolation? Maybe, but I think outputting the normals at the vertices already provides good information.

Reply:
No these ones aren't actually interpolated IIRC. We have concrete nodal normals at non-vertex nodes I'm pretty sure... my memory is so bad. yea ok it looks like I'm right
```c++
std::vector<Point>
AutomaticMortarGeneration::getNodalNormals(const Elem & secondary_elem) const
{
  std::vector<Point> nodal_normals(secondary_elem.n_nodes());
  for (const auto n : make_range(secondary_elem.n_nodes()))
    nodal_normals[n] = _secondary_node_to_nodal_normal.at(secondary_elem.node_ptr(n));

  return nodal_normals;
}
```
We loop over all the nodes here, not just vertices

Reply:
I see, didn't understand at first. We would need to make quite a few changes to this method to handle more general cases, e.g. select the interpolation order for output. But, yeah, you could see how potentially we would want to check geometry information for non-vertex nodes when applicable.

Reply:
can do in the future as desired